src/Bicep.Core/Registry/OciArtifactRegistry.cs

```diff
@@ -84,13 +84,15 @@ public override bool IsArtifactRestoreRequired(OciArtifactReference reference)
              * when we need to invalidate the cache, the module directory (or even a single file) should be deleted from the cache
              */
 
-            // TODO: Provider artifacts don't write a ModuleMain file, so this code is incorrect.
-            //   That can be solved by only writing layer data files only (see code under features.PublishSourceEnabled)
-            //   and not main.json directly (https://github.com/Azure/bicep/issues/11900)
-            return
-                !this.FileResolver.FileExists(this.GetModuleFileUri(reference, ModuleFileType.ModuleMain)) ||
-                !this.FileResolver.FileExists(this.GetModuleFileUri(reference, ModuleFileType.Manifest)) ||
-                !this.FileResolver.FileExists(this.GetModuleFileUri(reference, ModuleFileType.Metadata));
+            var artifactFilesNotFound = reference.Type switch
+            {
+                ArtifactType.Module => !this.FileResolver.FileExists(this.GetArtifactFileUri(reference, ArtifactFileType.ModuleMain)),
+                ArtifactType.Provider => !this.FileResolver.FileExists(this.GetArtifactFileUri(reference, ArtifactFileType.Provider)),
+                _ => default // should never happen
```

Comment:
[nit]: You want to fail early instead of returning the potentially wrong answer if the switch is no longer exhaustive. This would be a good place to [`throw new UnreachableException()`](https://github.com/dotnet/runtime/issues/35324).